### PR TITLE
add normalize transform

### PR DESCRIPTION
**Pull request by vince62s**
also fix a bug in the prefix apply_reverse

### CODE REVIEW

File: onmt/train_single.py [hunk 1]
```diff
@@ -23,6 +23,16 @@
...
+
+    # if transform + options set in 'valid' we need to copy in main
+    # transform / options for scoring considered as inference
+    validset_transforms = opt.data.get("valid", {}).get("transforms", None)
+    if validset_transforms:
+        opt.transforms = validset_transforms
+        if opt.data.get("valid", {}).get("tgt_prefix", None):
+            opt.tgt_prefix = opt.data.get("valid", {}).get("tgt_prefix", None)
+        if opt.data.get("valid", {}).get("src_prefix", None):
+            opt.src_prefix = opt.data.get("valid", {}).get("src_prefix", None)
```
Comment (francoishernandez) [inline onmt/train_single.py:35]: Suffix as well?

File: onmt/train_single.py [hunk 2]
```diff
@@ -67,6 +77,7 @@ def _init_train(opt):
             if len(old_transf) != 0:
                 _msg += f" -{old_transf}."
             logger.warning(_msg)
+            vocabs, transforms_cls = prepare_transforms_vocabs(opt)
```
Comment (francoishernandez) [inline onmt/train_single.py:80]: Not sure why this was not done before. Also, since it's a duplicate of line 83, and 86, it can probably be put only once after the conditions.

File: onmt/transforms/normalize.py [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+
+from itertools import chain
+
+from onmt.transforms import register_transform
+from .transform import Transform
+
+
+class MosesPunctNormalizer:
+    """
+    This is a Python port of the Moses punctuation normalizer from
```
[the remark this answers is not in the document]
Comment (francoishernandez) [reply onmt/transforms/normalize.py:15]: Just saw the ref line 214. Maybe cleaner to put it only once in the header of this file.

File: onmt/utils/optimizers.py [hunk 4]
```diff
@@ -259,7 +259,8 @@ def from_opt(cls, model, opt, checkpoint=None):
         optim_opt = opt
         optim_state_dict = None
 
-        if opt.train_from and checkpoint is not None:
+        if (opt.train_from and checkpoint is not None and
+                'optim' in checkpoint.keys()):
```
Comment (francoishernandez) [inline onmt/utils/optimizers.py:263]: I'm not sure this is a good idea. We could technically want to run this code with a released checkpoint (model['optim'] = None) and reset_optim = all.

Comment (vince62s) [reply onmt/utils/optimizers.py:263]: that's the point, you can't. The typical use case was nllb-200 checkpoint with no optimizer in it. it has to recreate the optimizer from scratch.

File: onmt/train_single.py (hunk 2 above)
Comment (vince62s) [reply onmt/train_single.py:80]: TBH even with this, the logic is not 100% bullet proof. But it helps. I will re-order things.

Comment (vince62s) [reply onmt/train_single.py:80]: Well there is a case:
if you want to do a plain train_from and just take the vocab options from the checkpoint, since the model_opt is loaded after, you need to keep the vocab from the checkpoint and not go through the prepare_transforms_vocabs(opt).
This is why I am saying the logic is not 100% bullet proof. I added the line in the case of transforms changes but technically it also requires the opt to define the vocab.

File: onmt/utils/optimizers.py (hunk 4 above)
Comment (vince62s) [reply onmt/utils/optimizers.py:263]: to be more specific.
case 'optim' key does not exist: we need to create the optimizer from scratch and the previous codebase would spit an error.
case 'optim' exist but checkpoint['optim'] = None: it goes through the code as it was.

Comment (francoishernandez) [reply onmt/utils/optimizers.py:263]: Makes sense, read it the other way around;
